File: pypp_cli/test_dir/cpp/src/built_in_functions/number_conversions.cpp
#include "built_in_functions/number_conversions.h"
#include "py_str.h"
#include "pypp_util/create/cstdint.h"
#include "pypp_util/create/others.h"
#include "pypp_util/print.h"
#include <cstdint>

namespace me {
void number_conversions_fn() {
    pypp::print(pypp::PyStr("NUMBER CONVERSIONS RESULTS:"));
    pypp::PyStr a = pypp::PyStr("123");
    pypp::PyStr b = pypp::PyStr("45.67");
    int a_int = pypp::int_(a);
    double b_float = pypp::float_(b);
    float b_float32 = pypp::to_float32(b_float);
    bool a_bool = pypp::bool_(a_int);
    pypp::print(pypp::PyStr(
        std::format("int(a): {}, float(b): {}, float32(b): {}, bool(a): {}",
                    a_int, b_float, b_float32, a_bool)));
    double a_float = pypp::float_(a_int);
    float a_float32 = pypp::to_float32(a_float);
    int b_int = pypp::int_(b_float);
    bool b_bool = pypp::bool_(b_float);
    pypp::print(pypp::PyStr(
        std::format("float(a): {}, float32(a): {}, int(b): {}, bool(b): {}",
                    a_float, a_float32, b_int, b_bool)));
    int8_t a_int8_t = pypp::to_int8_t(257);
    int16_t a_int16_t = pypp::to_int16_t(a_int);
    int32_t a_int32_t = pypp::to_int32_t(a_int);
    int64_t a_int64_t = pypp::to_int64_t(a_int);
    uint8_t a_uint8_t = pypp::to_uint8_t(a_int);
    uint16_t a_uint16_t = pypp::to_uint16_t(a_int);
    uint32_t a_uint32_t = pypp::to_uint32_t(a_int);
    uint64_t a_uint64_t = pypp::to_uint64_t(a_int);
    pypp::print(pypp::PyStr(
        std::format("int8_t: {}, int16_t: {}, int32_t: {}, int64_t: {}, "
                    "uint8_t: {}, uint16_t: {}, uint32_t: {}, uint64_t: {}",
                    a_int8_t, a_int16_t, a_int32_t, a_int64_t, a_uint8_t,
                    a_uint16_t, a_uint32_t, a_uint64_t)));
}

} // namespace me

File: pypp_cli/test_dir/python/src/built_in_functions/number_conversions.py
from pypp_python import (
    float32,
    to_float32,
    int8_t,
    int16_t,
    int32_t,
    int64_t,
    uint8_t,
    uint16_t,
    uint32_t,
    uint64_t,
    to_int8_t,
    to_int16_t,
    to_int32_t,
    to_int64_t,
    to_uint8_t,
    to_uint16_t,
    to_uint32_t,
    to_uint64_t,
)


def number_conversions_fn():
    print("NUMBER CONVERSIONS RESULTS:")
    a: str = "123"
    b: str = "45.67"

    # converting str to numbers
    a_int: int = int(a)
    b_float: float = float(b)
    b_float32: float32 = to_float32(b_float)
    a_bool: bool = bool(a_int)  # should be True since a_int is 123
    print(
        f"int(a): {a_int}, float(b): {b_float}, "
        f"float32(b): {b_float32}, bool(a): {a_bool}"
    )

    # converting numbers of different types
    a_float: float = float(a_int)
    a_float32: float32 = to_float32(a_float)
    b_int: int = int(b_float)
    b_bool: bool = bool(b_float)
    print(
        f"float(a): {a_float}, float32(a): {a_float32}, int(b): {b_int}, "
        f"bool(b): {b_bool}"
    )

    # converting cstdint numbers of different types
    a_int8_t: int8_t = to_int8_t(257)  # This is too large for int8_t. Goes to like 1
    a_int16_t: int16_t = to_int16_t(a_int)
    a_int32_t: int32_t = to_int32_t(a_int)
    a_int64_t: int64_t = to_int64_t(a_int)
    a_uint8_t: uint8_t = to_uint8_t(a_int)
    a_uint16_t: uint16_t = to_uint16_t(a_int)
    a_uint32_t: uint32_t = to_uint32_t(a_int)
    a_uint64_t: uint64_t = to_uint64_t(a_int)
    print(
        f"int8_t: {a_int8_t}, int16_t: {a_int16_t}, int32_t: {a_int32_t}, "
        f"int64_t: {a_int64_t}, uint8_t: {a_uint8_t}, uint16_t: {a_uint16_t}, "
        f"uint32_t: {a_uint32_t}, uint64_t: {a_uint64_t}"
    )
